devito/function.py

```diff
@@ -362,30 +365,33 @@ def __init__(self, *args, **kwargs):
         if not self._cached():
             super(TimeFunction, self).__init__(*args, **kwargs)
             self.time_dim = kwargs.get('time_dim', None)
+            if self.time_dim is not None and not isinstance(self.time_dim, TimeDimension):
+                raise ValueError("time_dim must be a TimeDimension, not %s" %
+                                 type(self.time_dim))
             self.time_order = kwargs.get('time_order', 1)
-            self.save = kwargs.get('save', False)
-
-            if not self.save:
-                if self.time_dim is not None:
-                    warning('Explicit time dimension size (time_dim) found for '
-                            'TimeFunction symbol %s, despite \nusing a stepping time '
-                            'dimension (save=False). This value will be ignored!'
-                            % self.name)
-                self.time_dim = self.time_order + 1
-                self.indices[0].modulo = self.time_dim
+            self.save = kwargs.get('save', None)
+
+            if self.save is not None:
+                if not isinstance(self.save, int):
+                    raise ValueError("save must be an int indicating the number of " +
+                                     "timesteps to be saved (is %s)" % type(self.save))
+                available_mem = virtual_memory().available
+
+                if np.dtype(self.dtype).itemsize * self.save > available_mem:
+                    warning("Trying to allocate more memory for symbol %s " % self.name +
+                            "than available on physical device, this will start swapping")
+                self.time_size = self.save
             else:
-                if self.time_dim is None:
-                    error('Time dimension (time_dim) is required'
-                          'to save intermediate data with save=True')
-                    raise ValueError("Unknown time dimensions")
+                self.time_size = self.time_order + 1
+                self.indices[0].modulo = self.time_size
```

Comment:
Is this changing the state of a ``Dimension``? If so, this is probably super dangerous. I know that's what we had before too, I just never noticed it.

Reply:
Oh this is not even the biggest problematic case of changing state on "global" dimension objects - the time reversal in the Operator is more dangerous than this. Outside the scope of this PR though.

Reply:
Agreed. One last comment: the change of state in ``Operator`` can probably be easily avoided (we can just have an instance-level flag and add a parameter to ``Iteration`` to allow describing things such as the traversal direction etc. On the other hand, I'm unsure how I would fix this.